Add "Disappearing messages" documentation (#133)
xmtp/docs-xmtp-org: docs/pages/inboxes/disappearing-messages.md

Disappearing messages are messages that are intended to be visible to users for only a short period of time. After the message expiration time passes, the messages are removed from the UI and deleted from local storage so the messages are no longer accessible to conversation participants.

Disappearing message behavior is enforced by apps, meaning that apps are responsible for removing messages from their UIs and local storage based on conditions set at the conversation level. Conversation participants using apps that support disappearing messages will have a UX that honors the message expiration conditions. 

However, it's important to understand that:

- A conversation participant using an app that doesn't support disappearing messages will not experience disappearing message behavior.
- Messages aren't deleted from the XMTP network.

Therefore, disappearing messages should be understood as best-effort, app-level privacy that helps avoid leaving an easily accessible record in a messaging UI. However, it's not a guaranteed, system-wide “message self-destruct.”

## Implement disappearing messages

### Enable disappearing messages for a conversation

> When creating or updating a conversation, enable users with appropriate permissions to set disappearing message expiration conditions.

Reviewer: for now only admins can set/remove it. and this policy is fixed.

- `from_ns`: Starting timestamp from which the message lifespan is calculated
- `in_ns`: Duration of time during which the message should remain visible to conversation participants

For example:

1. Set `from_ns` to the current time, such as `1738620126404999936` (nanoseconds since the Unix epoch of January 1, 1970).
2. Set `in_ns` to the message lifespan, such as 1800000000000000 (30 minutes).
3. Use `from_ns` and `in_ns` to calculate the message expiration time of `1738620126404999936 + 1800000000000000 = 1740420126404999936`.

To learn more see [conversation.rs](https://github.com/xmtp/libxmtp/blob/main/bindings_node/src/conversation.rs#L49).

### Send a disappearing message

When sending a message, it abides by message expiration conditions set for the conversation. For example:

```text
// Example: disappearing message with an expiration time based on a 
// countdown start time and expiration countdown
```

### Automatic deletion from local storage

A background worker runs every one second to clean up expired disappearing messages. The worker automatically deletes expired messages from local storage. No additional action is required by integrators.

> 

Author: not sure if we need a code sample for this -- unless we want to show how the worker is identifying expired messages?

Reviewer: hmm, not much to show here b/c they don't need to do anything about it, but still if we need to show some codes how things work here, we can put a link to the worker.

Disappearing messages are messages that are intended to be visible to users for only a short period of time. After the message expiration time passes, the messages are removed from the UI and deleted from local storage so the messages are no longer accessible to conversation participants.

Disappearing message behavior is enforced by apps, meaning that apps are responsible for removing messages from their UIs and local storage based on conditions set at the conversation level. Conversation participants using apps that support disappearing messages will have a UX that honors the message expiration conditions. 

However, it's important to understand that:

- A conversation participant using an app that doesn't support disappearing messages won't experience disappearing message behavior.
- Messages aren't deleted from the XMTP network.

:::tip

Therefore, disappearing messages should be understood as best-effort, app-level privacy that helps avoid leaving an easily accessible record in a messaging UI. However, it's not a guaranteed, system-wide “message self-destruct.”

:::

## Implement disappearing messages

### Enable disappearing messages for a conversation

> When creating or updating a conversation, admins can set disappearing message expiration conditions.

Author: @mchenani Did I capture your feedback correctly here? Specifically, only admins can set this?

Would it be helpful to even more specifically say:

When creating or updating a conversation, only group admins and DM participants can set disappearing message expiration conditions.

Reviewer: That's all correct, and I like the second suggestion since it's covering the DMs as well.

> When creating or updating a conversation, admins can set disappearing message expiration conditions.

Author: ```suggestion
When creating or updating a conversation, only group admins and DM participants can set disappearing message expiration conditions.
```

- `in_ns`: Duration of time during which the message should remain visible to conversation participants

For example:

1. Set `from_ns` to the current time, such as `1738620126404999936` (nanoseconds since the Unix epoch of January 1, 1970).
2. Set `in_ns` to the message lifespan, such as 1800000000000000 (30 minutes).
3. Use `from_ns` and `in_ns` to calculate the message expiration time of `1738620126404999936 + 1800000000000000 = 1740420126404999936`.

To learn more see [conversation.rs](https://github.com/xmtp/libxmtp/blob/main/bindings_node/src/conversation.rs#L49).

### Send a disappearing message

When sending a message, it abides by message expiration conditions set for the conversation. For example:

:::code-group

```tsx [Browser]
// add code sample
```

```tsx [Node]
// add code sample
```

```tsx [React Native]
// add code sample
```

```kotlin [Kotlin]
// add code sample
```

```swift [Swift]
// add code sample
```

:::

### Automatic deletion from local storage

A background worker runs every one second to clean up expired disappearing messages. The worker automatically deletes expired messages from local storage. No additional action is required by integrators.

> To learn more about the background worker, see [link to background worker code](#todo).

Author: @mchenani thank you for the suggestion - yes, it would be great to link to relevant background worker code - could you suggest a link, please?

Reviewer: https://github.com/xmtp/libxmtp/blob/main/xmtp_mls/src/groups/disappearing_messages.rs#L68
This is the worker code, also you can link the whole file, since it only contains the disappearing messages cleaner worker.

> To learn more about the background worker, see [link to background worker code](#todo).

Author: ```suggestion
To learn more about the background worker, see [disappearing_messages.rs](https://github.com/xmtp/libxmtp/blob/main/xmtp_mls/src/groups/disappearing_messages.rs#L68).
```

Disappearing messages are messages that are intended to be visible to users for only a short period of time. After the message expiration time passes, the messages are removed from the UI and deleted from local storage so the messages are no longer accessible to conversation participants.

Disappearing message behavior is enforced by apps, meaning that apps are responsible for removing messages from their UIs and local storage based on conditions set at the conversation level. Conversation participants using apps that support disappearing messages will have a UX that honors the message expiration conditions. 

However, it's important to understand that:

- A conversation participant using an app that doesn't support disappearing messages will not experience disappearing message behavior.
- Messages aren't deleted from the XMTP network.

Therefore, disappearing messages should be understood as best-effort, app-level privacy that helps avoid leaving an easily accessible record in a messaging UI. However, it's not a guaranteed, system-wide “message self-destruct.”

## **Implement disappearing messages**

Here is a high-level view of the disappearing messages flow:

> 1. When a 1:1 or group chat conversation is created or updated, a user with appropriate permissions can set disappearing message conditions for messages sent in the conversation. This includes:

Reviewer: Yes.

   - `from_ns`: Starting timestamp from which the message lifespan is calculated
   - `in_ns`: Duration for which the message should remain visible to conversation participants

> 2. In a conversation with disappearing message conditions, a sender creates a message. The message abides by disappearing message metadata set for the conversation.

Author: Just to be sure I understand - while the conditions are set at the conversation level, the message does not include metadata that indicates the conditions -- rather the dev would get that info by looking at the parent conversation, is that right?

Reviewer: Yea, that's correct.

# Support disappearing messages with XMTP

> Disappearing messages are messages that are intended to be visible to users for only a short period of time. After the message expiration time passes, the messages are removed from the UI and deleted from local storage so the messages are no longer accessible to conversation participants.

Reviewer: > the messages are removed from the UI

I would say we can't guarantee this, we delete the message from the local db, but if the UI doesn't refresh on-time the message can be visible till the next fresh!

Reviewer: Oh I see in the next paragraph you've mentioned that is the app's responsibility! ty🙌

Disappearing message behavior is enforced by apps, meaning that apps are responsible for removing messages from their UIs and local storage based on conditions set at the conversation level. Conversation participants using apps that support disappearing messages will have a UX that honors the message expiration conditions. 

However, it's important to understand that:

- A conversation participant using an app that doesn't support disappearing messages will not experience disappearing message behavior.

> - Messages aren't deleted from the XMTP network.

Reviewer: Great point! Thanks

2. In a conversation with disappearing message conditions, a sender creates a message. The message abides by disappearing message metadata set for the conversation.
3. The message is sent over the XMTP network.
4. The recipients’ apps with disappearing message support receive the message and check the expiration conditions set for the conversation.
5. If the message hasn't expired, the app displays the message in the messaging UI.
6. If the message has expired, the message is removed from (or not displayed in the first place in) the conversation participants’ app UIs and is deleted from the apps’ local storage.

### Enable disappearing messages for a conversation

When creating or updating a conversation, enable users with appropriate permissions to set disappearing message expiration conditions.

This includes setting the following conditions expressed in nanoseconds (ns):

- `from_ns`: Starting timestamp from which the message lifespan is calculated
- `in_ns`: Duration of time during which the message should remain visible to conversation participants

> For example:

Author: Did I get this correct, @mchenani? 🫣 😆

Reviewer 2: I'd say lets wait for the specific code snippets till it's out in the sdks since things will be differently exposed.

1. Set `from_ns` to the current time, such as `1738620126404999936` (nanoseconds since the Unix epoch of January 1, 1970).
2. Set `in_ns` to the message lifespan, such as 1800000000000000 (30 minutes).
3. Use `from_ns` and `in_ns` to calculate the message expiration time of `1738620126404999936 + 1800000000000000 = 1740420126404999936`.

To learn more see [conversation.rs](https://github.com/xmtp/libxmtp/blob/main/bindings_node/src/conversation.rs#L49).

### Send a disappearing message

When sending a message, it abides by message expiration conditions set for the conversation. For example:

```text
// Example: disappearing message with an expiration time based on a 
// countdown start time and expiration countdown
```

### Receive a disappearing message

On the receiving side, your app needs to:

1. Check for disappearing message metadata set for the conversation. For example:
    
    ```text
    // Example: check for disappearing message on the message's parent conversation
    ```
    

> 2. Display the message if the current time is before the expiration time. For example:

Reviewer: actually, we don't do that, receiving and processing a message happens in a separate flow, and we let the flow continue it's work as before, here the worker delete the messages that expired.

    
    ```text
    // Example: Check is the current time is before the expiration time
    ```
    
    Here are some tips for displaying disappearing messages in your app to ensure that users understand which messages use the feature and what to expect.
    
    - **Use a distinct visual style**: You might style disappearing messages differently from regular messages (e.g., a different background color or icon) so users understand they're short-lived.
    - **Display a countdown**: Show a small countdown that indicates how much time remains before the message disappears.

### Periodically check message expiration time

> A worker runs every two seconds to clean up expired disappearing messages. The worker also updates the message expiration policy to align with any updated policy values. For example:

Reviewer: ```suggestion
A worker runs every one second to clean up expired disappearing messages. The worker also updates the message expiration policy to align with any updated policy values. For example:
```

we changed it to 1 second.

```text

> // Example: Use the worker that runs every two seconds 

Reviewer: ```suggestion
// Example: Use the worker that runs every one second
```

> // to clean up expired messages based on group disappearing settings. 

Reviewer: ```suggestion
// to clean up expired messages based on conversation(group/dm) disappearing settings. 
```

> // to clean up expired messages based on conversation(group/dm) disappearing settings. 

Author: ```suggestion
// to clean up expired messages based on conversation (group/dm) disappearing settings. 
```

> // The worker will also update the message expiration policy to align 

Reviewer: ```suggestion
```
Tbh the worker does not do this, the worker only checks the messages if needed to be deleted, the settings will be received via a commit that includes the conversation's update info.

// with any updated disappearing message policy values.
```

### Remove a message from the app UI

> If the expiration time has passed, remove the message from the app UI. For example:

Reviewer: We don't do this, but maybe @nplasterer knows better, I assume we do have a flow in kotlin for example that will refresh when the database changes. is that correct?

Reviewer 2: Things are not cached so if they are removed from the local database the next time a call is made to the local database the messages should not be present.

```text
// Example: remove the message from your UI
```

### Delete a message from local storage

If the expiration time has passed, delete the message from local storage. For example:

```text

> // Example: delete the message from local storage

Reviewer: I would say there is no example to show here, because the integrator doesn't need to do anything about it, that would happen automatically by the worker.
